packages/account/src/providers/provider.ts

```diff
@@ -495,10 +514,87 @@ export default class Provider {
         fullRequest = await options.requestMiddleware(fullRequest);
       }
 
-      return options.fetch ? options.fetch(url, fullRequest, options) : fetch(url, fullRequest);
+      const operationName = fullRequest.body
+        ?.toString()
+        .match(/"operationName":"(.+)"/)?.[1] as keyof SdkOperations;
+
+      if (BLOCK_HEIGHT_SENSITIVE_OPERATIONS.includes(operationName)) {
+        const currentBlockHeight = this.currentBlockHeightCache[url.replace(/-sub$/, '')] ?? 0;
+
+        fullRequest.body = fullRequest.body
+          ?.toString()
+          .replace(/}$/, `,"extensions":{"required_fuel_block_height":${currentBlockHeight}}}`);
+      }
+
+      return Provider.fetchAndProcessBlockHeight(url, fullRequest, options);
     }, retryOptions);
   }
 
+  private static async fetchAndProcessBlockHeight(
+    url: string,
+    request: RequestInit,
+    options: ProviderOptions
+  ): Promise<Response> {
+    let response: Response;
+    let blockHeightPreconditionFailed = false;
+
+    const retryOptions: RetryOptions = {
+      maxRetries: 5,
+      baseDelay: 500,
+    };
+
+    let retryAttempt = 0;
+
+    do {
+      response = await (options.fetch ? options.fetch(url, request, options) : fetch(url, request));
+
+      const responseClone = response.clone();
+
+      let extensions: {
+        current_fuel_block_height?: number;
+        fuel_block_height_precondition_failed: boolean;
+      };
+
+      if (url.endsWith('-sub')) {
+        const reader = responseClone.body?.getReader() as ReadableStreamDefaultReader<Uint8Array>;
+        const { event } = await FuelGraphqlSubscriber.readEvent(reader);
+
+        extensions = event?.extensions as typeof extensions;
+      } else {
+        extensions = (await responseClone.json()).extensions;
+      }
+
+      Provider.setCurrentBlockHeight(
+        url.replace(/-sub$/, ''),
+        extensions?.current_fuel_block_height
+      );
```

Comment:
Can `Provider.setCurrentBlockHeight` handle the url replace?

Reply:
https://github.com/FuelLabs/fuels-ts/pull/3741/commits/1d16d3ad5e156121f91581a7b4060876e54df41b